Release/Growl/GrowlPreInstall/GrowlPreInstall.1
.\"Modified from man(1) of FreeBSD, the NetBSD mdoc.template, and mdoc.samples.
.\"See Also:
.\"man mdoc.samples for a complete listing of options
.\"man mdoc for the short list of editing options
.\"/usr/share/misc/mdoc.template
.Dd 9/10/07               \" DATE 
.Dt GrowlPostInstall 1      \" Program name and manual section number 
.Os Darwin
.Sh NAME                 \" Section Header - required - don't modify 
.Nm GrowlPostInstall,
.\" The following lines are read in generating the apropos(man -k) database. Use only key
.\" words here as the database is built based on the words here and in the .ND line. 
.Nm Other_name_for_same_program(),
.Nm Yet another name for the same program.
.\" Use .Nm macro to designate other names for the documented program.
.Nd This line parsed for whatis database.
.Sh SYNOPSIS             \" Section Header - required - don't modify
.Nm
.Op Fl abcd              \" [-abcd]
.Op Fl a Ar path         \" [-a path] 
.Op Ar file              \" [file]
.Op Ar                   \" [file ...]
.Ar arg0                 \" Underlined argument - use .Ar anywhere to underline
arg2 ...                 \" Arguments
.Sh DESCRIPTION          \" Section Header - required - don't modify
Use the .Nm macro to refer to your program throughout the man page like such:
.Nm
Underlining is accomplished with the .Ar macro like this:
.Ar underlined text .
.Pp                      \" Inserts a space
A list of items with descriptions:
.Bl -tag -width -indent  \" Begins a tagged list 
.It item a               \" Each item preceded by .It macro
Description of item a
.It item b
Description of item b
.El                      \" Ends the list
.Pp
A list of flags and their descriptions:
.Bl -tag -width -indent  \" Differs from above in tag removed 
.It Fl a                 \"-a flag as a list item
Description of -a flag
.It Fl b
Description of -b flag
.El                      \" Ends the list
.Pp
.\" .Sh ENVIRONMENT      \" May not be needed
.\" .Bl -tag -width "ENV_VAR_1" -indent \" ENV_VAR_1 is width of the string ENV_VAR_1
.\" .It Ev ENV_VAR_1
.\" Description of ENV_VAR_1
.\" .It Ev ENV_VAR_2
.\" Description of ENV_VAR_2
.\" .El                      
.Sh FILES                \" File used or created by the topic of the man page
.Bl -tag -width "/Users/joeuser/Library/really_long_file_name" -compact
.It Pa /usr/share/file_name
FILE_1 description
.It Pa /Users/joeuser/Library/really_long_file_name
FILE_2 description
.El                      \" Ends the list
.\" .Sh DIAGNOSTICS       \" May not be needed
.\" .Bl -diag
.\" .It Diagnostic Tag
.\" Diagnostic informtion here.
.\" .It Diagnostic Tag
.\" Diagnostic informtion here.
.\" .El
.Sh SEE ALSO 
.\" List links in ascending order by section, alphabetically within a section.
.\" Please do not reference files that do not exist without filing a bug report
.Xr a 1 , 
.Xr b 1 ,
.Xr c 1 ,
.Xr a 2 ,
.Xr b 2 ,
.Xr a 3 ,
.Xr b 3 
.\" .Sh BUGS              \" Document known, unremedied bugs 
.\" .Sh HISTORY           \" Document history if command behaves in a unique manner

changes.txt
0.7

Growl Core
- Added a status item (#23)
- Growl sends notifications when an applications registers (#112)
- Passive version check (#62)
- Added an option to log all notifications (#27)
- Squelch mode (#26)
- Networking support
- Bonjour support (#21)

Preference Pane
- Ability to enable or disable click handlers (#143)
- New about tab (#139)
- Application search field (#41)
- New preview button (#32)

Display Plugins
- New display: WebKit (#2). See http://growl.info/documentation/webkit.php for more information.
- New display: Brushed (#5)
- Bezel: new option for fading/shrinking (#170)
- Bezel: two new variants (#30)
- More customizable colors (#86, #34)
- MusicVideo: new effect when a queued notification fades in (#125)
- Mouse over effect for Bezel, Brushed, Bubbles and Smoke (#144)
- Hold non sticky notifications on screen with mouseover (#133)
- Multimonitor support (#87)
- Smaller notifications without title or description (#56)
- Duration sliders (#54)
- Opacity sliders (#44, #74)

Extras
- Removed growliChat, it has moved to http://www.growlichat.com
- New automatic summary option for GrowlMail
- Fixed summary mode (#100)
- GrowlMail is now compatible with Mail 2.0 (Tiger)
- Fixed GrowlMail image fetching from Address Book (#198)
- Fixed GrowlTunes showing 5 stars for unrated tracks (#127)
- GrowlSafari now sends different notifications for short and long downloads (#106)
- GrowlSafari can now send notifications for RSS feed updates
- growlnotify can wait until the notification is clicked or timed out (#42)

Developers and Scripters
- Growl Core
	- Passive registration (#70)
	- Coalescing support (#33)
	- Get GrowlHelperApp version number via the command line (#29)
	- Applications can now specify a preferred display plugin (#126)
- Framework
	- Carbon apps can have a Growl installer now (#77)
	- Fixed click notifiations using the Carbon API (#152)
	- Easier functions for registering manually (#91)
	- Only return click & timeout notification to the application requesting it (#203)
- Bindings
	- Sticky support for the Java bindings (#72)
	- Icon support for the Perl bindings (#6)
- Documentation
	- More sample code (#82)


0.6.2

General
- Resolved an issue where the version update button would crash System Preferences. See ticket http://trac.growl.info/trac/ticket/8
- Smoke was defaulting to 0 percent opacity, making it unreadable. Resolved this issue, and defaulted it to 75%. See ticket http://trac.growl.info/trac/ticket/79

Developers and Scripters
- Carbon
	- Resolved an issue where Growl_SetDelegate would return success even if the registrationDictionary is not filled out. Thanks to Kent Sorensen (Snak) for finding this problem. See ticket http://trac.growl.info/trac/ticket/80
- Perl
	- Resolved problems with the Perl bindings in regards to Mac::Glue tests. See ticket http://trac.growl.info/trac/ticket/47
	- Updated the Perl bindings. See ticket http://trac.growl.info/trac/ticket/53 for more information.


0.6.1

- Resolved issue with the framework installed with Growl. Thanks to Timothy Hatcher to finding this issue.
- Updated Entourage Advanced script to work with Growl .6.x

0.6
New to the DMG
	- Scripts
	- Extras
	- More documentation

Bindings (available in a separate download on the website)
	- Java
	- Ruby

Prefpane
	- General Tab
		- Check version button

	- Applications Tab
		- Priority
		- Sticky
		- Remove button

	- Network Tab
New Displays
	- Smoke
	- Music Video
	- NSLog
	- MailMe
	- Speech

Changes to current displays:
	Bubbles:
	- Color changing abilities
	- Support for priority (color change)

	Bezel:
	- Support for priority (brings higher priority items to the front of the queue)

New Localizations
	- German

0.5 - Release
Prefpane
	- General Tab
	- Applications Tab
	- Display Options

New Displays
	- Bubbles
	- Bezel

Bindings
	- AppleScript
	- Carbon
	- Cocoa
	- Perl
	- Python
	- TCL
	


Release/Growl/GrowlPreInstall/GrowlPreInstall.m
#import <Cocoa/Cocoa.h>
#import "AEVTBuilder.h"

@interface NSWorkspace (ProcessSerialNumberFinder)
- (ProcessSerialNumber)processSerialNumberForApplicationWithIdentifier:(NSString *)identifier;
@end

@implementation NSWorkspace (ProcessSerialNumberFinder)
- (ProcessSerialNumber)processSerialNumberForApplicationWithIdentifier:(NSString *)identifier
{
	ProcessSerialNumber psn = {0, 0};

	NSEnumerator *enumerator = [[self launchedApplications] objectEnumerator];
	NSDictionary *dict;
	while ((dict = [enumerator nextObject])) {
		if ([[dict objectForKey:@"NSApplicationBundleIdentifier"] isEqualToString:identifier]) {
			psn.highLongOfPSN = [[dict objectForKey:@"NSApplicationProcessSerialNumberHigh"] longValue];
			psn.lowLongOfPSN  = [[dict objectForKey:@"NSApplicationProcessSerialNumberLow"] longValue];
			break;
		}
	}
	
	return psn;
}
@end

int main (int argc, const char * argv[]) {
    NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];

	NSAppleEventDescriptor *descriptor;
	/* tell application "System Preferences" to quit. The name may be localized, so we can't use applescript directly. */
	descriptor = [AEVT class:kCoreEventClass id:kAEQuitApplication
					  target:[[NSWorkspace sharedWorkspace] processSerialNumberForApplicationWithIdentifier:@"com.apple.systempreferences"],
				  ENDRECORD];
	[descriptor sendWithImmediateReplyWithTimeout:5];

	/* tell application "GrowlHelperApp" to quit */
	[NSTask launchedTaskWithLaunchPath:@"/usr/bin/killall" arguments:[NSArray arrayWithObject:@"GrowlHelperApp"]];
	[NSTask launchedTaskWithLaunchPath:@"/usr/bin/killall" arguments:[NSArray arrayWithObject:@"GrowlMenu"]];

	/* delete old Growl installations */
	[[NSFileManager defaultManager] removeFileAtPath:[[[NSHomeDirectory()
														stringByAppendingPathComponent:@"Library"]
													   stringByAppendingPathComponent:@"PreferencePanes"] 
													  stringByAppendingPathComponent:@"Growl.prefPane"] handler:nil];

	/* We'll be running sudo'd from the installer, so this will have appropriate permissions */
	[[NSFileManager defaultManager] removeFileAtPath:@"/Library/PreferencePanes/Growl.prefPane" handler:nil];
	

    [pool release];
    return 0;
}
